swift/ql/lib/codeql/swift/elements/decl/EnumDecl.qll

```diff
@@ -1,4 +1,51 @@
-// generated by codegen/codegen.py, remove this comment if you wish to edit this file
 private import codeql.swift.generated.decl.EnumDecl
+private import codeql.swift.elements.decl.EnumCaseDecl
+private import codeql.swift.elements.decl.EnumElementDecl
+private import codeql.swift.elements.decl.Decl
 
-class EnumDecl extends Generated::EnumDecl { }
+/**
+ * An enumeration declaration, for example:
+ * ```
+ * enum MyColours {
+ *   case red
+ *   case green
+ *   case blue
+ * }
+ * ```
+ */
+class EnumDecl extends Generated::EnumDecl {
+  /**
+   * Gets the number of `EnumElementDecl`s in this enumeration before the `memberIndex`th member. Some
+   * of the members of an `EnumDecl` are `EnumCaseDecls` (representing the `case` lines), each of
+   * which holds one or more `EnumElementDecl`s.
+   */
+  private int countEnumElementsTo(int memberIndex) {
+    memberIndex = 0 and result = 0
+    or
+    exists(Decl prev | prev = this.getMember(memberIndex - 1) |
+      result = this.countEnumElementsTo(memberIndex - 1) + prev.(EnumCaseDecl).getNumberOfElements()
+      or
+      not prev instanceof EnumCaseDecl and
+      result = this.countEnumElementsTo(memberIndex - 1)
+    )
+  }
+
+  /**
+   * Gets the `index`th enumeration element of this enumeration (0-based).
+   */
+  final EnumElementDecl getEnumElement(int index) {
+    exists(int memberIndex |
+      result =
+        this.getMember(memberIndex)
+            .(EnumCaseDecl)
+            .getElement(index - this.countEnumElementsTo(memberIndex))
+    )
+  }
+
+  /**
+   * Gets an enumeration element of this enumeration.
+   */
+  final EnumElementDecl getAnEnumElement() {
+    result = this.getMember(_).(EnumCaseDecl).getElement(_)
+  }
```

Comment:
Two questions:
1. For an `EnumDecl d` we now have both `d.getAMember()` and `d.getAnEnumElement()`. Should we override `getAMember()` in this case to add QLDoc that mention the difference between this new predicate and `getAMember`?

2. Now that we're making it more difficult for users to accidentally get hold of an `EnumCaseDecl`, how about we completely get rid of them from the AST by replacing `any()` with `none()` in https://github.com/github/codeql/blob/main/swift/ql/lib/codeql/swift/elements/decl/EnumCaseDeclConstructor.qll#L4?

Reply:
1. Good idea, clearer documentation is always a good thing.
2. I'm a little reluctant to get rid of the ability to reason about `EnumCaseDecl`s entirely.  The fashion at the moment is very much for security queries where `EnumCaseDecl` is a useless detail.  But some users may want to write style queries - for example "only put one enum value in each enum case declaration" would be a valid style query.

Reply:
Good point about 2. Let's leave in those `EnumCaseDecl`s for now.

Reply:
Back on 1., unfortunately I can't override the generated `getMember` and `getAMember` methods as they are marked `final`.  Let me know if you can think of a workaround that will work for documentation purposes.  I'm tempted to put a general warning on `Decl.getMember` itself (the ordered version) to prefer not to rely on the order of elements via this predicate when alternative methods are available?

Reply:
Ah, I wasn't aware that they're marked as `final`. Maybe they should't be for exactly this reason 🤔. I can't think of any good workaround for this either.

> I'm tempted to put a general warning on Decl.getMember itself (the ordered version) to prefer not to rely on the order of elements via this predicate when alternative methods are available?

Sounds good to me.

cc @redsun82 maybe we should remove the `final` annotation on these predicates (so that we can override them with custom QLDoc).

Reply:
FYI, `Decl.getMember/0` is final because it's a generated accessor; the current extension point would be `Decl.getImmediateMember/0`.

Another alternative is to make the `EnumCaseDecl`s hidden, i.e. they resolve to their child `EnumElementDecl`s.

Reply:
It's `Decl.getMember/1` that's problematic, because the indexes aren't what you might expect (if you didn't realize the `case`s are children interleaves with the actual elements).  It works fine, it's just potentially confusing.

> the current extension point would be Decl.getImmediateMember/0.

Yeah, you're right, thanks.  Though I'm really just looking to update the QLDoc / help text on `getMember/1` for `EnumDecl`, which I've achieved by just rewording the root version.

Reply:
> Another alternative is to make the `EnumCaseDecl`s hidden, i.e. they resolve to their child `EnumElementDecl`s.

This would only work if each `EnumCaseDecl` would have exactly one `EnumElementDecl`, which is not the case.

Btw, with https://github.com/github/codeql/pull/13249 the customization point would be `Decl.getMember`, as `Decl`s would cease to be hideable.